src/buildHTML.js

```diff
@@ -88,146 +64,173 @@ export const buildExpression = function(
-    // spans for determining spacings between surrounding atoms.
-    const nonSpaces: (?HtmlDomNode)[] = [
-        surrounding[0] ? makeSpan([surrounding[0]], [], options) : null,
-        ...rawGroups.filter(group => group && group.classes[0] !== "mspace"),
-        surrounding[1] ? makeSpan([surrounding[1]], [], options) : null,
-    ];
+    // Dummy spans for determining spacings between surrounding atoms.
+    // If `expression` has no atoms on the left or right, class "leftmost"
+    // or "rightmost", respectively, is used to indicate it.
+    const dummyPrev = makeSpan([surrounding[0] || "leftmost"], [], options);
+    const dummyNext = makeSpan([surrounding[1] || "rightmost"], [], options);
+
+    // TODO: These code assumes that a node's math class is the first element
+    // of its `classes` array. A later cleanup should ensure this, for
+    // instance by changing the signature of `makeSpan`.
 
     // Before determining what spaces to insert, perform bin cancellation.
     // Binary operators change to ordinary symbols in some contexts.
-    for (let i = 1; i < nonSpaces.length - 1; i++) {
-        const nonSpacesI: HtmlDomNode = assert(nonSpaces[i]);
-        const left = getOutermostNode(nonSpacesI, "left");
-        if (left.classes[0] === "mbin" &&
-                isBinLeftCanceller(nonSpaces[i - 1], isRealGroup)) {
-            left.classes[0] = "mord";
+    traverseNodes(groups, (node, prev) => {
+        const prevType = prev.classes[0];
+        const type = node.classes[0];
+        if (prevType === "mbin" && utils.contains(binRightCanceller, type)) {
+            prev.classes[0] = "mord";
+        } else if (type === "mbin" && utils.contains(binLeftCanceller, prevType)) {
+            node.classes[0] = "mord";
         }
-
-        const right = getOutermostNode(nonSpacesI, "right");
-        if (right.classes[0] === "mbin" &&
-                isBinRightCanceller(nonSpaces[i + 1], isRealGroup)) {
-            right.classes[0] = "mord";
+    }, {node: dummyPrev}, dummyNext);
+
+    traverseNodes(groups, (node, prev) => {
+        const prevType = getTypeOfDomTree(prev);
+        const type = getTypeOfDomTree(node);
+
+        // 'mtight' indicates that the node is script or scriptscript style.
+        const space = prevType && type ? (node.hasClass("mtight")
+            ? tightSpacings[prevType][type]
+            : spacings[prevType][type]) : null;
+        if (space) { // Insert glue (spacing) after the `prev`.
+            return buildCommon.makeGlue(space, glueOptions);
         }
+    }, {node: dummyPrev}, dummyNext);
+
+    return groups;
+};
+
+// Depth-first traverse non-space `nodes`, calling `callback` with the current and
+// previous node as arguments, optionally returning a node to insert after the
+// previous node. `prev` is an object with the previous node and `insertAfter`
+// function to insert after it. `next` is a node that will be added to the right.
+const traverseNodes = function(
```

Comment:
It might be worth unit testing this.  Also the name is very general but it some logic that's super specific to modifying spacing.  Can you change the name and update the comment to reflect this?

Reply:
@kevinbarabash I've renamed the function to `traverseNonSpaceNodes` and added comments about its usage. We really need a way to unit-test functions.